Project Walking/Source code/app/src/main/java/com/teamapricot/projectwalking/LocationHandler.java

```diff
@@ -0,0 +1,151 @@
+package com.teamapricot.projectwalking;
+
+import android.Manifest;
+import android.location.LocationListener;
+import android.location.LocationManager;
+
+import androidx.appcompat.app.AppCompatActivity;
+
+import com.teamapricot.projectwalking.dialogs.PermissionRejectedDialog;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * @author Daniel Brännvall
+ * @version 2021-09-20
+ *
+ * A class for accessing location data.
+ */
+public class LocationHandler {
+    private final String TAG = "LocationHandler";
+
+    private final PermissionHandler permissionHandler;
+
+    private Position position = null;
+
+    private final LocationManager locationManager;
+    private LocationListener locationListener = null;
+
+    private final int updateInterval;
+
+    private final List<UpdateListener> updateListeners;
+
+    /**
+     * Constructor.
+     *
+     * @param activity The associated Activity
+     * @param interval Interval (in ms) between location updates
+     */
+    LocationHandler(AppCompatActivity activity, int interval) {
+        permissionHandler = new PermissionHandler(activity);
+        locationManager = (LocationManager) activity.getApplicationContext()
+                .getSystemService(AppCompatActivity.LOCATION_SERVICE);
+        updateListeners = new ArrayList<>();
+        updateInterval = interval;
+
+        if (permissionHandler.checkPermission(Manifest.permission.ACCESS_FINE_LOCATION)) {
+            initializeLocationHandler();
+        } else {
+            permissionHandler.requestPermissionAsync(Manifest.permission.ACCESS_FINE_LOCATION)
+                    .thenAccept(permissionGranted -> {
+                        if (permissionGranted) {
+                            initializeLocationHandler();
+                        } else {
```

Comment:
This call to `initializeLocationHandler` does not work properly when I test it. I.e. if you withdraw the location permission from within settings, then restart the app, it will ask for permissions as it's supposed to. However, `thenAccept` runs on its separate permissionHandler thread, not the calling thread. If I accept the permission, the TextView does not update its values. I have to restart the app so that the permission is already granted when the app starts for it to work properly. It seems like `initializeLocationHandler` needs to execute on the UI thread, otherwise the position update will not work properly in this case. Wrap the call to `initializeLocationHandler` in a call to `activity.runOnUiThread`. This seems to solve the issue, or wrap the line(s) of code inside `initializeLocationHandler` to make sure that they always run on the UI thread. Your choice 😄

Reply:
The settings kept crashing on the virtual Pixel 4 I was testing it on, and I guess I got distracted and didn't test thoroughly. Sorry!

Yeah, it's probably best to run it on the UI thread, added the wrapper.